### PR TITLE
drivers/w1: add 1-wire shell

### PR DESCRIPTION
**Pull request by str4t0m**
This PR implements shell support for 1-Wire.
Commands for bus-reset, bit-, byte-, and block- transfer,
as well as search and configuration are implemented.

### CODE REVIEW

File: drivers/w1/w1_shell.c [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+/* 1-Wire reset bus <device> */
+static int cmd_w1_reset_bus(const struct shell *sh, size_t argc, char **argv)
+{
+	const struct device *dev;
+	int ret;
+
+	dev = device_get_binding(argv[1]);
+	if (!dev) {
```
Comment (ycsin) [inline drivers/w1/w1_shell.c:58]: is additional `device_is_ready` check necessary?

Comment (str4t0m) [reply drivers/w1/w1_shell.c:58]: Compared to `DEVICE_DT_GET`, `device_get_binding` only returns a pointer if the device was successfully initialized.

File: drivers/w1/w1_shell.c [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+		shell_error(sh, W1DEV_X_NOT_FOUND, argv[1]);
+		return -EINVAL;
+	}
+
+	(void)w1_lock_bus(dev);
+	ret = w1_read_bit(dev);
+	if (ret < 0) {
+		shell_error(sh, "Failed to read bit [%d]", ret);
+	} else {
+		shell_print(sh, "Output: 0b%x", ret);
```
[the remark this answers is not in the document]
Comment (str4t0m) [reply drivers/w1/w1_shell.c:90]: This can only be 0, or 1, therefore I decided to prefix it with 0b.

**Comment by str4t0m** [comment]
Not sure why the PDF Documentation build was canceled.
Let's see if it works this time.

**Comment by str4t0m** [comment]
Rebased to resolve merge conflict.

**Comment by dp7hgh7** [comment]
This module requires the label to be set in devicetree (if I understand correctly). Since this is currently fading from the tree, it might be a good idea to mention that in the help text.

File: drivers/w1/w1_shell.c [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+	}
+
+	(void)w1_unlock_bus(dev);
+	return ret;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_w1,
+	SHELL_CMD_ARG(reset, NULL,
+		     "Reset 1-Wire bus.\n"
+		     "<device_name>",
```
Comment (dp7hgh7) [inline drivers/w1/w1_shell.c:375]:
```suggestion
		     "<device_label>",
```
This makes it more obvious that this is the `label` field and not the node name.

File: drivers/w1/w1_shell.c [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+		shell_error(sh, "Failed to initiate search [%d]", ret);
+	} else {
+		shell_print(sh, "Found %d device(s)", ret);
+	}
+
+	(void)w1_unlock_bus(dev);
+	return ret;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_w1,
```
Comment (dp7hgh7) [inline drivers/w1/w1_shell.c:372]: This section breaks with auto formatting. I suggest to exclude it from clang-format.

**Review by str4t0m — COMMENTED** [review]
Comments should be addressed. Could you please have another look.

File: drivers/w1/w1_shell.c (hunk 4 above)
Comment (dp7hgh7) [reply drivers/w1/w1_shell.c:374]: Fine for me

**Comment by dp7hgh7** [comment]
> The device is looked up via device_get_binding, therefore matches the node's label property if it exists, and otherwise the nodes full name. Therefore I'd rather not change it to device_label.
Checking other shell implementations, many seem to use <device>. Should I short it to simly <device>?

Makes sense to me. The relevant thing is that it is something different than `DEVICE_DT_GET(...)` consumes.

**Comment by str4t0m** [comment]
The 1-Wire shell is now enabled `default y if !SHELL_MINIMAL`.

File: drivers/w1/w1_shell.c (hunk 3 above)
Comment (str4t0m) [reply drivers/w1/w1_shell.c]: changed <device_name> to <device> in the help text.

**Comment by str4t0m** [comment]
The indentation of long sub-command help texts was not correctly displayed, therefore I have removed the indentation for all subcmds.

**Review by casparfriedrich — DISMISSED** [review]

**Review by casparfriedrich — APPROVED** [review]

**Review by ycsin — APPROVED** [review]

**Pull request merged**